packages/core/src/dispatcher.ts

```diff
@@ -63,12 +71,42 @@ export default class Dispatcher extends EventEmitter {
    */
   async register (document: Document): Promise<void> {
     this._documents[document.id.toString()] = document
-    // request tip
-    const payload = { typ: MsgType.REQUEST, id: document.id.toString(), doctype: document.doctype.doctype }
+
+    // Build a QUERY message to send to the pub/sub topic to request the latest tip for this document
+    const payload = await this._buildQueryMessage(document)
+
+    // Store the query id so we'll process the corresponding RESPONSE message when it comes in
+    this._outstandingQueryIds[payload.id] = document.id.baseID
+
     this._ipfs.pubsub.publish(this.topic, JSON.stringify(payload))
     this._log({ peer: this._peerId, event: 'published', topic: this.topic, message: payload })
   }
 
+  async _buildQueryMessage(document: Document): Promise<Record<string, any>> {
+    const message = { typ: MsgType.QUERY, doc: document.id.baseID.toString() }
+
+    // Add 'id' to message that is a hash of the message contents.
+    const id = await this._hashMessage(message)
+
+    return {...message, id: id.toString()}
+  }
+
+  /**
+   * Computes a sha-256 multihash of the input message canonicalized using dag-cbor
+   * @param message
+   */
+  async _hashMessage(message: any) : Promise<Uint8Array> {
+    // DAG-CBOR encoding
+    let id: Uint8Array = dagCBOR.util.serialize(message)
+
+    // SHA-256 hash
+    id = sha256.hash(id)
+
+    // Multihash encoding
+    const buf = Buffer.from(id)
+    return multihashes.encode(buf, 'sha2-256')
```

Comment:
multihashes doesn't support Uint8Arrays?

Reply:
Doesn't look like it, I get a compilation error if I try to pass in the Uint8Array directly ☹️